sphinxcontrib/datatemplates/directive.py

```diff
@@ -26,39 +27,67 @@ class DataTemplate(rst.Directive):
         'template': rst.directives.unchanged,
         'csvheaders': rst.directives.flag,
         'csvdialect': _handle_dialect_option,
+        'encoding': rst.directives.unchanged,
     }
     has_content = False
 
-    def _load_data(self, env, data_source):
+    def _load_csv(self, filename, encoding):
+        try:
+            if encoding is None:
+                f = open(filename, 'r', newline='')
+            else:
+                f = codecs.open(filename, 'r', newline='', encoding=encoding)
+            dialect = self.options.get('csvdialect')
+            if dialect == "auto":
+                sample = f.read(8192)
+                f.seek(0)
+                sniffer = csv.Sniffer()
+                dialect = sniffer.sniff(sample)
+            if 'csvheaders' in self.options:
+                if dialect is None:
+                    r = csv.DictReader(f)
+                else:
+                    r = csv.DictReader(f, dialect=dialect)
+            else:
+                if dialect is None:
+                    r = csv.reader(f)
+                else:
+                    r = csv.reader(f, dialect=dialect)
+            return list(r)
+        finally:
+            f.close()
+
+    def _load_json(self, filename, encoding):
+        try:
+            if encoding is None:
+                f = open(filename, 'r')
+            else:
+                f = codecs.open(filename, 'r', encoding=encoding)
+            return json.load(f)
+        finally:
+            f.close()
+
+    def _load_yaml(self, filename, encoding):
+        try:
+            if encoding is None:
+                f = open(filename, 'r')
+            else:
+                f = codecs.open(filename, 'r', encoding=encoding)
+            return yaml.safe_load(f)
+        finally:
+            f.close()
+
+    def _load_data(self, env, data_source, encoding):
```

Comment:
We could have 1 block in _load_data() to manage opening the file, and then pass a file handle to the other load functions. That would cut down on the number of places we have to pay attention to the encoding.

Reply:
Actually I think so, too - but open for CSV does take different parameters and for XML there is no open at all.

Reply:
Maybe it's better to tackle #17 first.

Reply:
Yeah, for #17 I could see refactoring the data loading to a method of the subclass and then it wouldn't feel as repetitive.